src/jit/importer.cpp

```diff
@@ -3932,12 +3932,9 @@ GenTree* Compiler::impIntrinsic(GenTree*                newobjThis,
         }
     }
 
-    if (mustExpand)
+    if (mustExpand && (retNode == nullptr))
     {
-        if (retNode == nullptr)
-        {
-            NO_WAY("JIT must expand the intrinsic!");
-        }
+        return impUnsupportedIntrinsic(CORINFO_HELP_THROW_NOT_IMPLEMENTED, method, sig, mustExpand);
```

Comment:
Does this mean that issues like this #10215 and this #14401 will no longer be compile time failures, but get silently transformed by the JIT to runtime failures?

Reply:
This just causes the JIT to throw a `NotImplementedException` in the case we have a `mustExpand` intrinsic that was unhandled by the runtime. This case should, realistically, never be hit.

However, when bringing up larger intrinsic sets (such as for the Hardware Intrinsic work), we often need to expose the methods, mirror those in CoreFX, and then implement them in CoreCLR which temporarily exposes some APIs that will be unimplemented. This can cause issues for things like the PMI jobs (as in #20757).

Reply:
Ah, I think I see. The concern is that this will mask the case where the JIT hits some bug that prevented the expansion of an intrinsic (such as was the case for `InitializeArray` due to a stack spill). Is that correct?

@CarolEidt, should we have some mechanism so that these can still be exposed as JIT asserts, rather than only as runtime failures (which might mask a case if it is on a non-executed code-path)?

Reply:
How do we end up with mustExpand = true for the unimplemented intrinsics here?

The case that I'm worried about is when `getIntrinsicID` sets this flag (i.e. the runtime says we need to expand) and based on my reading of the code, RyuJIT ignores it and generates bad code instead (a throw).

There were RyuJIT bugs in the past (the above mentioned pull requests) that were caught by the `NO_WAY` assert above that will now basically require executing the generating code path and tracing the failure back to the compilation phase to root cause. Seems like a lot more work than just hitting an assert.

Reply:
It might be useful to have an assert for this. As @tannergooding pointed out, the reason we run into these is that, while we're adding new intrinsics, it is not feasible to keep the JIT and the log in perfect sync.

Reply:
Would it make sense to replace the bool with a 3 state variable? "No opinion about expanding", "expand or generate throwing code", and "expand or fail"? I see on the runtime side that if we say mustExpand in `getIntrinsicID`, we really mean it.

Reply:
> How do we end up with mustExpand = true for the unimplemented intrinsics here?

In CoreCLR, `mustExpand` is true for any JIT Intrinsic which is recursive to itself (which are effectively all HardwareIntrinsics. For example: https://source.dot.net/#System.Private.CoreLib/shared/System/Runtime/Intrinsics/X86/Sse.cs,fd01f10f9e63dcb4). This condition is set here: https://github.com/dotnet/coreclr/blob/4d32755ab8336e305e0e84d867334d1af957cd92/src/jit/importer.cpp#L3380

> The case that I'm worried about is when getIntrinsicID sets this flag (i.e. the runtime says we need to expand) and based on my reading of the code, RyuJIT ignores it and generates bad code instead (a throw).

Right, this changes the behavior so that an intrinsic that fails to expand will throw an exception, rather than fail fast.

Reply:
I'd suggest we do this conversion to throw only for the hardware intrinsics.

For other sorts of intrinsics the implementation in the jit usually happens before or at the same time the intrinsic is exposed, since some kind of implementation is needed to ensure that the intrinsic is viable.

Reply:
Updated to only throw NYI for hardware intrinsics.